### DISCUSSION TITLE
Math rendering: expression limit?

### DISCUSSION THREAD

**Discussion by stevenvergenz** [community/community#45276] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Bug

### Body

I am attempting to display a complicated matrix in a Markdown file, but it refuses to render, reporting "Unable to render expression". All the pieces individually work fine, but putting it together fails. Is there some character limit on evaluated math expressions?

This works: (rows 1 and 2 of my matrix)

$$
\begin{bmatrix}
{\color{red}a} {\color{teal}r} + {\color{red}b} {\color{teal}u} + {\color{red}c} {\color{teal}x}
& {\color{red}a} {\color{orange}s} + {\color{red}b} {\color{orange}v} + {\color{red}c} {\color{orange}y}
& {\color{red}a} {\color{magenta}t} + {\color{red}b} {\color{magenta}w} + {\color{red}c} {\color{magenta}z}
\\\\
{\color{green}d} {\color{teal}r} + {\color{green}e} {\color{teal}u} + {\color{green}f} {\color{teal}x}
& {\color{green}d} {\color{orange}s} + {\color{green}e} {\color{orange}v} + {\color{green}f} {\color{orange}y}
& {\color{green}d} {\color{magenta}t} + {\color{green}e} {\color{magenta}w} + {\color{green}f} {\color{magenta}z}
\end{bmatrix}
$$

This works too: (rows 2 and 3 of my matrix)

$$
\begin{bmatrix}
{\color{green}d} {\color{teal}r} + {\color{green}e} {\color{teal}u} + {\color{green}f} {\color{teal}x}
& {\color{green}d} {\color{orange}s} + {\color{green}e} {\color{orange}v} + {\color{green}f} {\color{orange}y}
& {\color{green}d} {\color{magenta}t} + {\color{green}e} {\color{magenta}w} + {\color{green}f} {\color{magenta}z}
\\\\
{\color{lightblue}g} {\color{teal}r} + {\color{lightblue}h} {\color{teal}u} + {\color{lightblue}i} {\color{teal}x}
& {\color{lightblue}g} {\color{orange}s} + {\color{lightblue}h} {\color{orange}v} + {\color{lightblue}i} {\color{orange}y}
& {\color{lightblue}g} {\color{magenta}t} + {\color{lightblue}h} {\color{magenta}w} + {\color{lightblue}i} {\color{magenta}z}
\end{bmatrix}
$$

But when I put them together:

$$
\begin{bmatrix}
{\color{red}a} {\color{teal}r} + {\color{red}b} {\color{teal}u} + {\color{red}c} {\color{teal}x}
& {\color{red}a} {\color{orange}s} + {\color{red}b} {\color{orange}v} + {\color{red}c} {\color{orange}y}
& {\color{red}a} {\color{magenta}t} + {\color{red}b} {\color{magenta}w} + {\color{red}c} {\color{magenta}z}
\\\\
{\color{green}d} {\color{teal}r} + {\color{green}e} {\color{teal}u} + {\color{green}f} {\color{teal}x}
& {\color{green}d} {\color{orange}s} + {\color{green}e} {\color{orange}v} + {\color{green}f} {\color{orange}y}
& {\color{green}d} {\color{magenta}t} + {\color{green}e} {\color{magenta}w} + {\color{green}f} {\color{magenta}z}
\\\\
{\color{lightblue}g} {\color{teal}r} + {\color{lightblue}h} {\color{teal}u} + {\color{lightblue}i} {\color{teal}x}
& {\color{lightblue}g} {\color{orange}s} + {\color{lightblue}h} {\color{orange}v} + {\color{lightblue}i} {\color{orange}y}
& {\color{lightblue}g} {\color{magenta}t} + {\color{lightblue}h} {\color{magenta}w} + {\color{lightblue}i} {\color{magenta}z}
\end{bmatrix}
$$

Is this a bug in the renderer, or am I missing something? Thanks.

**Comment by jge162** [comment]
It is possible that the renderer is having trouble with the size of the expression that you are trying to render.  you should break up the expressions into smaller chunks and render them separately. Additionally, try removing any unnecessary elements such as colors or formatting that may be causing the expression to be too large. then test

**Reply by stevenvergenz** [reply]
It looks like it works when I remove some colors:

$$
\begin{bmatrix}
	  {a} {\color{teal}r}    + {b} {u}    + {c} {\color{teal}x}
	& {a} {\color{orange}s}  + {b} {v}  + {c} {\color{orange}y}
	& {a} {\color{magenta}t} + {b} {w} + {c} {\color{magenta}z}
	\\\\
	  {\color{green}d} {\color{teal}r}    + {\color{green}e} {\color{teal}u}    + {\color{green}f} {\color{teal}x}
	& {\color{green}d} {\color{orange}s}  + {\color{green}e} {\color{orange}v}  + {\color{green}f} {\color{orange}y}
	& {\color{green}d} {\color{magenta}t} + {\color{green}e} {\color{magenta}w} + {\color{green}f} {\color{magenta}z}
	\\\\
	  {\color{lightblue}g} {\color{teal}r}    + {\color{lightblue}h} {\color{teal}u}    + {\color{lightblue}i} {\color{teal}x}
	& {\color{lightblue}g} {\color{orange}s}  + {\color{lightblue}h} {\color{orange}v}  + {\color{lightblue}i} {\color{orange}y}
	& {\color{lightblue}g} {\color{magenta}t} + {\color{lightblue}h} {\color{magenta}w} + {\color{lightblue}i} {\color{magenta}z}
\end{bmatrix}
$$

That seems a little arbitrary. How is the system determining what expressions are "too large"? There's no character limit on code blocks here, why for math blocks?

**Reply by jge162** [reply]
prevent the system from crashing or slowing down due to excessively large or complex expressions.

**Reply by nschloe** [reply]
MathJax itself has no problems rendering the thing, so I would also think there's a bug.

**Comment by nschloe** [comment]
Turns out that this only depends on the `color`s. If you have 50 or more of them, rendering will fail. Added this to the bug list https://github.com/nschloe/github-math-bugs#50-or-more-colors-in-a-block.

**Reply by 1ec5** [reply]
Even one color in a comment can cause this error: https://github.com/osm-americana/openstreetmap-americana/pull/896#discussion_r1257865900.

**Reply by nschloe** [reply]
@1ec5 It's not the color that's not working in your example, but the `\rule`.

**Reply by 1ec5** [reply]
Thanks, yes, I just realized that and replaced it with a simple `\blacksquare`.

**Comment by tueda** [comment]
I also ran into this size limit. It would be better if it were explicitly stated somewhere.

As of August 2024, the first of the following two examples is correctly rendered, but the second is not (the standard MathJax renders both).

```math
\rlap{\kern0pt\raise3pt.}\rlap{\kern1pt\raise2pt.}\rlap{\kern2pt\raise1pt.}\rlap{\kern3pt\raise0pt.}\rlap{\kern5pt\raise0pt.}\rlap{\kern6pt\raise1pt.}\rlap{\kern6pt\raise2pt.}\rlap{\kern7pt\raise2pt.}\rlap{\kern7pt\raise3pt.}\rlap{\kern7pt\raise4pt.}\rlap{\kern8pt\raise4pt.}\rlap{\kern8pt\raise5pt.}\rlap{\kern8pt\raise6pt.}\rlap{\kern9pt\raise7pt.}\rlap{\kern9pt\raise8pt.}\rlap{\kern9pt\raise9pt.}\rlap{\kern10pt\raise10pt.}\rlap{\kern10pt\raise11pt.}\rlap{\kern10pt\raise12pt.}\rlap{\kern11pt\raise13pt.}\rlap{\kern11pt\raise14pt.}\rlap{\kern11pt\raise15pt.}\rlap{\kern12pt\raise15pt.}\rlap{\kern12pt\raise16pt.}\rlap{\kern12pt\raise17pt.}\rlap{\kern13pt\raise17pt.}\rlap{\kern13pt\raise18pt.}\rlap{\kern14pt\raise19pt.}\rlap{\kern15pt\raise19pt.}\rlap{\kern16pt\raise19pt.}\rlap{\kern16pt\raise18pt.}\rlap{\kern17pt\raise18pt.}\rlap{\kern17pt\raise17pt.}\rlap{\kern18pt\raise17pt.}\rlap{\kern18pt\raise16pt.}\rlap{\kern19pt\raise16pt.}\rlap{\kern19pt\raise15pt.}\rlap{\kern20pt\raise15pt.}\rlap{\kern20pt\raise16pt.}\rlap{\kern21pt\raise16pt.}\rlap{\kern22pt\raise17pt.}\rlap{\kern22pt\raise18pt.}\rlap{\kern23pt\raise19pt.}\rlap{\kern23pt\raise20pt.}\rlap{\kern24pt\raise21pt.}\rlap{\kern24pt\raise22pt.}\rlap{\kern24pt\raise23pt.}\rlap{\kern25pt\raise24pt.}\rlap{\kern25pt\raise25pt.}\rlap{\kern25pt\raise26pt.}\rlap{\kern26pt\raise27pt.}\rlap{\kern26pt\raise28pt.}\rlap{\kern26pt\raise29pt.}\rlap{\kern27pt\raise30pt.}\rlap{\kern27pt\raise31pt.}\rlap{\kern28pt\raise32pt.}\rlap{\kern28pt\raise33pt.}\rlap{\kern29pt\raise34pt.}\rlap{\kern30pt\raise34pt.}\rlap{\kern30pt\raise35pt.}\rlap{\kern31pt\raise35pt.}\rlap{\kern31pt\raise34pt.}\rlap{\kern32pt\raise34pt.}\rlap{\kern32pt\raise33pt.}\rlap{\kern33pt\raise33pt.}\rlap{\kern33pt\raise32pt.}\rlap{\kern34pt\raise32pt.}\rlap{\kern34pt\raise31pt.}\rlap{\kern35pt\raise31pt.}\rlap{\kern36pt\raise31pt.}\rlap{\kern37pt\raise32pt.}\rlap{\kern37pt\raise33pt.}\rlap{\kern38pt\raise33pt.}\rlap{\kern38pt\raise34pt.}\rlap{\kern38pt\raise35pt.}\rlap{\kern39pt\raise35pt.}\rlap{\kern39pt\raise36pt.}\rlap{\kern39pt\raise37pt.}\rlap{\kern40pt\raise38pt.}\rlap{\kern40pt\raise39pt.}\rlap{\kern40pt\raise40pt.}\rlap{\kern41pt\raise41pt.}\rlap{\kern41pt\raise42pt.}\rlap{\kern41pt\raise43pt.}\rlap{\kern42pt\raise44pt.}\rlap{\kern42pt\raise45pt.}\rlap{\kern42pt\raise46pt.}\rlap{\kern43pt\raise46pt.}\rlap{\kern43pt\raise47pt.}\rlap{\kern43pt\raise48pt.}\rlap{\kern44pt\raise48pt.}\rlap{\kern44pt\raise49pt.}\rlap{\kern45pt\raise50pt.}\rlap{\kern47pt\raise50pt.}\rlap{\kern48pt\raise49pt.}\rlap{\kern49pt\raise48pt.}\rlap{\kern50pt\raise47pt.}
```

```math
\rlap{\kern3pt\raise0pt.}\rlap{\kern3pt\raise1pt.}\rlap{\kern4pt\raise2pt.}\rlap{\kern4pt\raise3pt.}\rlap{\kern4pt\raise4pt.}\rlap{\kern4pt\raise5pt.}\rlap{\kern5pt\raise6pt.}\rlap{\kern5pt\raise7pt.}\rlap{\kern5pt\raise8pt.}\rlap{\kern5pt\raise9pt.}\rlap{\kern6pt\raise9pt.}\rlap{\kern6pt\raise10pt.}\rlap{\kern6pt\raise11pt.}\rlap{\kern7pt\raise12pt.}\rlap{\kern8pt\raise12pt.}\rlap{\kern8pt\raise11pt.}\rlap{\kern9pt\raise11pt.}\rlap{\kern9pt\raise10pt.}\rlap{\kern9pt\raise9pt.}\rlap{\kern10pt\raise9pt.}\rlap{\kern10pt\raise8pt.}\rlap{\kern10pt\raise7pt.}\rlap{\kern11pt\raise7pt.}\rlap{\kern12pt\raise7pt.}\rlap{\kern13pt\raise8pt.}\rlap{\kern13pt\raise9pt.}\rlap{\kern13pt\raise10pt.}\rlap{\kern14pt\raise11pt.}\rlap{\kern14pt\raise12pt.}\rlap{\kern14pt\raise13pt.}\rlap{\kern14pt\raise14pt.}\rlap{\kern15pt\raise15pt.}\rlap{\kern15pt\raise16pt.}\rlap{\kern15pt\raise17pt.}\rlap{\kern15pt\raise18pt.}\rlap{\kern16pt\raise19pt.}\rlap{\kern16pt\raise20pt.}\rlap{\kern16pt\raise21pt.}\rlap{\kern17pt\raise21pt.}\rlap{\kern17pt\raise22pt.}\rlap{\kern18pt\raise22pt.}\rlap{\kern19pt\raise22pt.}\rlap{\kern19pt\raise21pt.}\rlap{\kern20pt\raise20pt.}\rlap{\kern20pt\raise19pt.}\rlap{\kern21pt\raise18pt.}\rlap{\kern21pt\raise17pt.}\rlap{\kern22pt\raise17pt.}\rlap{\kern23pt\raise18pt.}\rlap{\kern23pt\raise19pt.}\rlap{\kern24pt\raise20pt.}\rlap{\kern24pt\raise21pt.}\rlap{\kern24pt\raise22pt.}\rlap{\kern24pt\raise23pt.}\rlap{\kern25pt\raise24pt.}\rlap{\kern25pt\raise26pt.}\rlap{\kern26pt\raise27pt.}\rlap{\kern26pt\raise28pt.}\rlap{\kern26pt\raise29pt.}\rlap{\kern26pt\raise30pt.}\rlap{\kern27pt\raise31pt.}\rlap{\kern27pt\raise32pt.}\rlap{\kern28pt\raise33pt.}\rlap{\kern29pt\raise33pt.}\rlap{\kern29pt\raise32pt.}\rlap{\kern30pt\raise31pt.}\rlap{\kern30pt\raise30pt.}\rlap{\kern31pt\raise29pt.}\rlap{\kern31pt\raise28pt.}\rlap{\kern32pt\raise28pt.}\rlap{\kern33pt\raise28pt.}\rlap{\kern33pt\raise29pt.}\rlap{\kern34pt\raise29pt.}\rlap{\kern34pt\raise30pt.}\rlap{\kern34pt\raise31pt.}\rlap{\kern35pt\raise32pt.}\rlap{\kern35pt\raise33pt.}\rlap{\kern35pt\raise34pt.}\rlap{\kern35pt\raise35pt.}\rlap{\kern36pt\raise36pt.}\rlap{\kern36pt\raise37pt.}\rlap{\kern36pt\raise38pt.}\rlap{\kern36pt\raise39pt.}\rlap{\kern37pt\raise40pt.}\rlap{\kern37pt\raise41pt.}\rlap{\kern37pt\raise42pt.}\rlap{\kern38pt\raise43pt.}\rlap{\kern39pt\raise43pt.}\rlap{\kern40pt\raise43pt.}\rlap{\kern40pt\raise42pt.}\rlap{\kern40pt\raise41pt.}\rlap{\kern41pt\raise41pt.}\rlap{\kern41pt\raise40pt.}\rlap{\kern41pt\raise39pt.}\rlap{\kern42pt\raise39pt.}\rlap{\kern42pt\raise38pt.}\rlap{\kern43pt\raise38pt.}\rlap{\kern44pt\raise39pt.}\rlap{\kern44pt\raise40pt.}\rlap{\kern44pt\raise41pt.}\rlap{\kern45pt\raise41pt.}\rlap{\kern45pt\raise42pt.}\rlap{\kern45pt\raise43pt.}\rlap{\kern45pt\raise44pt.}\rlap{\kern46pt\raise45pt.}\rlap{\kern46pt\raise46pt.}\rlap{\kern46pt\raise47pt.}\rlap{\kern46pt\raise48pt.}\rlap{\kern47pt\raise49pt.}\rlap{\kern47pt\raise50pt.}
```
